examples/pipelines/fl_cross_silo_factory/submit.py

```diff
@@ -282,132 +318,58 @@ def silo_scatter_subgraph(
         # Iteration number
         iteration_num=iteration_num,
     )
 
     # Assigning the silo's second compute to the training component
-    silo_training_step.compute = scatter_compute2
+    silo_training_step.compute = silo_compute2
 
-    # IMPORTANT: we will assume that any output provided here can be exfiltrated into the orchestrator
+    # IMPORTANT: we will assume that any output provided here can be exfiltrated into the orchestrator/gather
     return {
         # NOTE: the key you use is custom
         # a map function scatter_to_gather_map needs to be provided
         # to map the name here to the expected input from gather
-        "input_silo": silo_training_step.outputs.model
+        "model": silo_training_step.outputs.model
     }
 
 
-#######################
-### D. FACTORY CODE ###
-#######################
-
-# In this section, we're using a wrapper to build the full FL pipeline
-# based on the custom methods you implemented in section C.
+def get_scatter_constant_inputs():
+    """Get scatter constant inputs like lr, epochs, etc."""
+    return YAML_CONFIG.inputs
 
-# 1. create an instance of the factory
-builder = FederatedLearningPipelineFactory()
 
-# 2. feed it with FL parameters
+#######################
+### D. FL Contract ###
+#######################
 
-builder.set_orchestrator(
-    # provide settings for orchestrator
-    YAML_CONFIG.federated_learning.orchestrator.compute,
-    YAML_CONFIG.federated_learning.orchestrator.datastore,
-)
+from fl_helper import scatter_gather
 
-for silo_config in YAML_CONFIG.federated_learning.silos:
-    builder.add_silo(
-        # provide settings for this silo
-        silo_config.name,
-        silo_config.computes,
-        silo_config.datastore,
-        # any additional custom kwarg will be sent to silo_preprocessing() as is
-        raw_train_data=Input(
-            type=silo_config.training_data.type,
-            mode=silo_config.training_data.mode,
-            path=silo_config.training_data.path,
-        ),
-        raw_test_data=Input(
-            type=silo_config.testing_data.type,
-            mode=silo_config.testing_data.mode,
-            path=silo_config.testing_data.path,
-        ),
-    )
+scatter_configs = get_scatter_configs()
+gather_config = get_gather_config()
+scatter_constant_inputs = get_scatter_constant_inputs()
 
-# 3. use a pipeline factory method
-# 3.1 build flexible fl pipeline
-
-# DEBUG
-if args.debug:
-    logger = logging.getLogger()
-    logger.setLevel(logging.DEBUG)
-    log_format = logging.Formatter("[%(levelname)s] - %(message)s")
-    handler = logging.StreamHandler(sys.stdout)
-    handler.setLevel(logging.DEBUG)
-    handler.setFormatter(log_format)
-    logger.addHandler(handler)
-
-pipeline_job = builder.build_flexible_fl_pipeline(
-    # building the FL graph requires two arguments that can be either a component or a pipeline
-    # scatter is our subgraph above
+pipeline_job, fl_factory = scatter_gather(
```

Comment:
could we split the factory in two: 1) the scatter_gather() engine and 2) the soft validation engine, which could be in a completely distinct class (and import).

The goal would be to get the scatter_gather() helper very cleanly defined into a python script of its own, and the soft_validation as a completely distinct source code, that we can choose to pull in or not.

Reply:
Agree! done